### PR TITLE
Siva | BAH-1146 | Changes for form2 encounter validations

### PR DESCRIPTION
**Pull request by sivareddytw**
Approach :

- New End Point Created
- Reading form definition to get obs control properties (mandatory, allow future date, multi select and addmore)
- Validating mandatory and  allow future date properties.
- Creating multiple observations for multi select and add more obs.
- Using separate delimiters for multi select and add more obs which can be configured in global properties

### CODE REVIEW

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form1CSVObsHandler.java [hunk 1]
```diff
@@ -46,4 +46,9 @@ public Form1CSVObsHandler(CSVObservationHelper csvObservationHelper) {
         return observations;
     }
 
+    @Override
+    public List<EncounterTransaction.Observation> handle(EncounterRow encounterRow, boolean shouldPerformForm2Validations) throws ParseException {
+        return null;
```
Comment (gsluthra) [inline admin/src/main/java/org/bahmni/module/admin/observation/handler/Form1CSVObsHandler.java:51]: I am skeptical of "null" being returned. I am guessing there should be no case when a caller called this method on Form1CSVObsHandler. If so, then maybe it should throw a ParseException with message saying, called with form2validation on a Form1 handler. That will be better, than some chain ignoring this error, and instead ending up printing null / throwing a null pointer exception down the line. What do u think?  And, then u can add a test to say, if someone calls this method on Form1  handler, it should throw a ParseException.

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java [hunk 2]
```diff
@@ -57,6 +61,39 @@ public Form2CSVObsHandler(CSVObservationHelper csvObservationHelper, FormFieldPa
...
+                    setFormNamespaceAndFieldPath(form2Observations, form2CSVHeaderParts);
+                    validateObsForFutureDate(form2Observations, form2CSVObservation, form2CSVHeaderParts);
+                }
+            } else {
+                boolean mandatoryFieldMissing = formFieldPathService.isMandatory(form2CSVHeaderParts);
+                if(mandatoryFieldMissing) {
+                    throw new APIException(format("Empty value provided for mandatory field %s", form2CSVHeaderParts.get(form2CSVHeaderParts.size()-1)));
+                }
+            }
+        }
```
Comment (gsluthra) [inline admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java:93]: I feel this code could be made easier to read. 

Like verifyForMultiSelect(), verifyForAddMore(), verifyForMandatory(). And all the check and verification is embedded inside those methods. Right now this method has too many if-else-if-else in one go. Please try simplifying this. I feel some of the elses are also not needed. Some lines will return false automatically and don't need perf optimsation with if-else.

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java [hunk 3]
```diff
@@ -82,4 +119,75 @@ private void setFormNamespaceAndFieldPath(List<EncounterTransaction.Observation>
...
+        String path = tokens[formFieldPathPosition];
+        String controlIdPrefix = path.split("-")[0];
+
+        for(int i = 0; i < addmoreForm2CSVObservations.size(); i++) {
+            final EncounterTransaction.Observation observation = form2Observations.get(prevObsCount + i);
+            tokens[formFieldPathPosition] = controlIdPrefix + "-" + i;
+            observation.setFormFieldPath(String.join(FORM_FIELD_PATH_SEPARATOR, tokens));
+            observation.setFormNamespace(FORM_NAMESPACE);
+        }
+    }
```
Comment (gsluthra) [inline admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java:179]: This code is about parsing Form field locations -- and can be externalised into its own class with its own test cases -- and can be reused/extended independent of Form2CSVObsHandler. I would recommend extracting it for making current class easier to read, and have less logic.

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java (hunk 2 above)
Comment (sivareddytw) [reply admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java:81]: @gsluthra Thanks for review. I have refactored the method.

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java (hunk 3 above)
Comment (sivareddytw) [reply admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java]: Thanks Gurpreet for reviewing. I have refactored this to a new Class.

File: admin/src/main/java/org/bahmni/module/admin/observation/handler/Form1CSVObsHandler.java (hunk 1 above)
Comment (sivareddytw) [reply admin/src/main/java/org/bahmni/module/admin/observation/handler/Form1CSVObsHandler.java:51]: Thanks Gurpreet for reviewing this. This is a dummy method for form1Obs handler and will not executed from existing end point. This is called from ObservationMapper which contains list of Obs handlers. Returning null from this method and gracefully handling in the called method [ObservationMapper.getObservations(encounterRow,shouldPerformForm2Validations)]

**Comment by supriyatw** [comment]
Can we fix the build errors for this PR. @sivareddytw

File: admin/src/main/java/org/bahmni/module/admin/csv/service/FormFieldPathGeneratorService.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    public void setFormNamespaceAndFieldPathForAddmoreObs(List<EncounterTransaction.Observation> form2Observations, List<String> form2CSVHeaderParts, List<KeyValue> addmoreForm2CSVObservations) {
+        if (isEmpty(form2Observations)) {
+            return;
+        }
+        int prevObsCount = form2Observations.size() - addmoreForm2CSVObservations.size();
+        final String formFieldPath = formFieldPathService.getFormFieldPath(form2CSVHeaderParts);
+        String[] tokens = formFieldPath.split(FORM_FIELD_PATH_SEPARATOR);
+        int formFieldPathPosition = tokens.length;
+        formFieldPathPosition = tokens.length - 1;
```
[the remark this answers is not in the document]
Comment (sivareddytw) [reply admin/src/main/java/org/bahmni/module/admin/csv/service/FormFieldPathGeneratorService.java]: updated

**Comment by sivareddytw** [comment]
> Can we fix the build errors for this PR. @sivareddytw

It depends on changes from bahmni-java-utils/form2-utils which are yet to merge.

**Review by supriyatw — APPROVED** [review]

**Review by binduak — APPROVED** [review]

**Pull request merged**